File: overlay/typescript/hydra-kernel/src/main/typescript/hydra/overlay/typescript/lib/hashing.ts
// Hand-written runtime: hydra.lib.hashing primitives.
//
// SHA-256 over raw bytes (binary). In the TS runtime `binary` values are represented
// as base64-encoded strings (the convention from literals.ts / files.ts / text.ts), so
// these helpers decode the base64 input, hash it, and (for sha256) re-encode the digest
// as base64. Pure and total. Pairs with hydra.lib.files.readFile to hash file contents.
// Mirrors overlay/python/.../lib/hashing.py and Hydra.Lib.Hashing. For #524.

import { createHash } from "node:crypto";

// Compute the SHA-256 digest of binary (a base64-encoded string), returned as a
// base64-encoded string (binary).
export const sha256 = (data: string): string =>
  createHash("sha256").update(Buffer.from(data, "base64")).digest("base64");

// Compute the SHA-256 digest of binary (a base64-encoded string) as a 64-character
// lowercase hexadecimal string.
export const sha256Hex = (data: string): string =>
  createHash("sha256").update(Buffer.from(data, "base64")).digest("hex");

File: overlay/scala/hydra-kernel/src/main/scala/hydra/overlay/scala/lib/hashing.scala
package hydra.overlay.scala.lib

import java.security.MessageDigest

/**
 * Scala implementations of hydra.lib.hashing primitives (#524).
 *
 * The Hydra `binary` type maps to a base64-encoded String in the Scala host
 * (see literals.binaryToBytes), so these helpers base64-decode the input bytes,
 * hash them, and (for sha256) base64-encode the resulting digest. Pure and total.
 */
object hashing:
  private def digest(binary: String): Array[Byte] =
    val bytes = java.util.Base64.getDecoder.decode(binary)
    MessageDigest.getInstance("SHA-256").digest(bytes)

  /**
   * Compute the SHA-256 digest of binary (a base64-encoded String), returned as a
   * base64-encoded String (binary).
   */
  def sha256(binary: String): String =
    java.util.Base64.getEncoder.encodeToString(digest(binary))

  /**
   * Compute the SHA-256 digest of binary (a base64-encoded String) as a 64-character
   * lowercase hexadecimal string.
   */
  def sha256Hex(binary: String): String =
    digest(binary).map(b => f"${b & 0xff}%02x").mkString
